Duplicate Detection: REST Contract submission section & item link

kshepherd opened the pull request:
Adds REST contract descriptions for the Duplicate Detection feature, which has active PRs.

The initial work on this feature was supported by TU Berlin. This work was supported by FHNW and ZHAW.

## Related PRs:
Backend: https://github.com/DSpace/DSpace/pull/9265
Frontend: https://github.com/DSpace/dspace-angular/pull/2749
**Replaces**: https://github.com/DSpace/RestContract/pull/198

The changes are made to `items.md` (new item link) and `workspaceitems.md` (new submission section), with the example data provided in a new `workspaceitem-data-duplicates.md` file.

kshepherd commented on items.md:
```diff
@@ -569,6 +569,26 @@ Return codes:
 * 404 Not found - if the item doesn't exist
 
+### Get potential duplicates of an item
+**GET /api/core/items/{:item-uuid}/duplicates**
```
Thanks @tdonohue , I have pushed changes to refactor this endpoint to a searchBy method. I agree that an item link, in retrospect, is not the best fit.

kshepherd commented:
@tdonohue i have addressed all review feedback for this feature and @MW3000 has tested the dspace and angular features successfully. The PRs are still waiting on a second approval. Are we able to get another review in for this feature?

abollini commented on items.md:
```diff
@@ -586,3 +586,26 @@ Return codes:
+
+### findDuplicates
+
+**GET /api/core/items/search/findDuplicates?uuid=<:uuid>**
```
This endpoint is not aligned to our REST best practices. A search method in the items collection resource endpoint must return a list of Item Resource.
If we need this endpoint it should be moved to a new endpoint specific for duplicate groups but as the feature is intended to be used only in the submission I don't expect this to be really needed

kshepherd replied:
I have moved this to a separate controller

tdonohue replied:
@kshepherd : This REST Contract doesn't appear to be updated with your new endpoint/controller.  Could you update it please?

Also, take a look at my feedback on the backend PR, as I suggested you might rename this to be something like..
```
/api/submission/duplicates/search?uuid=
```

That is modeled similar to the Controlled vocabulary endpoints which are also used primarily in Submission: https://github.com/DSpace/RestContract/blob/main/vocabularies.md

abollini commented on workspaceitem-data-duplicates.md:
```diff
@@ -0,0 +1,89 @@
+# WorkspaceItem data of identifiers sectionType
```
I was surprised to found this endpoint as I expect a porting of the dspace-cris rest contract already in place
https://github.com/4Science/Rest7Contract/blob/main-cris/workspaceitem-data-detect-duplicate.md

tdonohue replied:
@abollini : While I understand your frustration that this feature is no longer a direct port from DSpace-CRIS, this new approach was clearly announced by @pnbecker back in January in all the prior PRs:
* https://github.com/DSpace/RestContract/pull/198#issuecomment-1899294323
* https://github.com/DSpace/DSpace/pull/8415#issuecomment-1898107717
* https://github.com/DSpace/dspace-angular/pull/1732#issuecomment-1898111164

I also remember this being discussed in a meeting in either Dec or January where @pnbecker noted they would need to refactor the approach if Library Code was still to take this work on.  Because this feature was deemed a priority for Steering, I said that sounded reasonable & no one raised objections in the meeting.

I was not aware of any objections myself to the new approach until you submitted your reviews.  The 4Science team has always been assigned as reviewers to these Duplicate Detection PRs as I wanted to get your feedback.

But at this point, I don't feel we can delay what is a "working feature" just because it doesn't align with DSpace-CRIS. This feedback about DSpace-CRIS alignment comes very far after our "Review/Test Deadline" (Feb 9), and it's now too late to address.

If, in the future, we decide the DSpace-CRIS implementation is a better approach...there is always the opportunity to migrate from one to the other (or similar) in DSpace 9.0 or later.  Just because one implementation is accepted, doesn't mean it cannot be replaced in a future release.

abollini reviewed (changes requested):
It is a pity to note that this feature is not longer a porting of the DSpace-CRIS detect duplicate feature available since 7.0 (actually version 4 more than 9 years ago).
The feature is much less extend than the one provided by DSpace-CRIS and look to be incompatible with this one. I see a risk in accepting this PR to make DSpace and DSpace-CRIS more far when instead we are trying to take the two projects as close as possible. I expect users to be confused by the differences and subtracting resources in the DSpace / DSpace-CRIS community that would be better spent in managing a convergence among these two projects.

kshepherd commented:
I have moved duplicates to a separate controller and documented it as such (see also: REST and angular PRs as linked in description). I have also shifted solr work so it just uses an existing non-tokenized field instead of introducing a _signature field to avoid confusion and conflict.
If the endpoint name should also be changed to avoid confusion or conflict with 4science work coming in the future let me know, the endpoint name/path is of no consequence to me.

tdonohue commented on submission.md:
```diff
@@ -66,3 +66,95 @@ The final Item's UUID will be the same as it was in the WorkspaceItem (i.e. the
+
+## Finding potential duplicate items
+
+**GET /api/submission/duplicates/search?uuid=<:uuid>**
```
I'd recommend moving this documentation to a new `duplicates.md` section.  This `submission.md` page is an **overview of how to perform a submission** from the REST API.  So, this doesn't belong here, as it's not a task you need to do to perform a submission from the UI.

Instead, this should have it's own "page" in the REST Contract, just like how the `/api/submission/vocabularies` endpoint (for controlled vocabs) has it's own page in `vocabularies.md`.

kshepherd replied:
No problem, done. also linked from endpoints.md

tdonohue reviewed (commented):
@kshepherd : Overall, this looks good.  I have a minor reorganization note though below that should be addressed.

kshepherd commented:
@tdonohue applied changes and rebased -- i added duplicates.md to the main section of endpoints.md since if this is merged, it will not be experimental/under discussion

abollini commented on duplicates.md:
```diff
@@ -0,0 +1,97 @@
+
+## Search
+
+**GET /api/submission/duplicates/search?uuid=<:uuid>**
```
this is not aligned to our REST architecture. We should have something like that
` /api/submission/duplicates/search/byUUID`

kshepherd replied:
yes. the angular data service will need updating too, if i am able to do this -- as i noted in the other PR comment, my understanding was that RestRepositories with searchBy methods etc require an addressable object, which the "duplicate" stubs are not (even though they are related to addressable Items).

tdonohue replied:
@kshepherd : I tried to clarify the RestRepository question in your [implementation PR here](https://github.com/DSpace/DSpace/pull/9265#discussion_r1509131702).  Please let us know if it still doesn't make sense though.

abollini reviewed (changes requested):
hi @kshepherd  the search method needs a small modification to conform to our practice.

kshepherd commented:
Pushed update with new notes about get single / get all not being implemented, and the new search path

tdonohue reviewed (approved):
👍 Thanks @kshepherd !  This looks good now.  Verified that all feedback has been addressed from this PR & the backend PR
